### PR TITLE
Fix regression: input arrays are not broadcast

### PR DESCRIPTION
**Pull request by gjoseph92**
I noticed that #49 introduced a regression where input arrays now all must be the same shape (before, they'd be broadcast when possible). In https://github.com/xgcm/xhistogram/pull/49#discussion_r640872588 I was trying to explain this and a potential fix, but probably wasn't very clear about it.

Here's a test demonstrating the issue; this passes on `HEAD^` (as long as dask is installed from main). I'll push up the rest of the fix in a few hours.

cc @TomNicholas @rabernat

### CODE REVIEW

**Comment by rabernat** [comment]
Thanks so much for this @gjoseph92! Sorry we missed your suggestion. #49 got a bit hard to follow by the end.

**Comment by TomNicholas** [comment]
Thanks @gjoseph92 for catching this!

The PR looks good to me, but there some test failures on some environments (all with python 3.7 it seems) of a form where
`all_arrays = broadcast_arrays(*all_arrays)`
is somehow causing
```
    def raise_if_computed():
>       raise ValueError("Triggered forbidden computation")
E       ValueError: Triggered forbidden computation

xhistogram\test\fixtures.py:8: ValueError
```

**Comment by TomNicholas** [comment]
> there are some test failures

Attempting to reproduce locally (on Ubuntu) suggests that a more recent numpy version will pass on python 3.7, so perhaps any future release of this functionality should just bump the dependencies a bit?

**Comment by gjoseph92** [comment]
Thanks for looking into that. Yes, this will require NumPy >= 0.17 so that [NEP-18](https://numpy.org/neps/nep-0018-array-function-protocol.html) is active by default. I've updated that in this PR. Incidentally I saw that the NumPy version used in CI is different for all three Python versions; is that intentional?

**Comment by jrbourbeau** [comment]
> I saw that the NumPy version used in CI is different for all three Python versions; is that intentional?

Yes, that's so we can provide some CI coverage over multiple versions of NumPy (we do something similar in `dask/dask`)

**Review by TomNicholas — APPROVED** [review]
Those pass now, and that explanation makes sense, thanks!

The only cross is from codecov - apparently the coverage somehow decreased slightly, but I think this fix can be merged anyway.
